File: leetcode/python/0205_isomorphic_strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t): return False
        match = {}

        for i in range(len(s)):
            if s[i] not in match:
                if t[i] in match.values(): return False
                match[s[i]] = t[i]
            else:
                if match[s[i]] != t[i]: return False
        
        return True

File: leetcode/cpp/0205_isomorphic_strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> letterMap;
        unordered_set<char> mappedToLetters;

        for (int i = 0; i < s.size(); i++) {
            auto itMap = letterMap.find(s[i]);

            if (itMap == letterMap.end()) {
                if (mappedToLetters.find(t[i]) != mappedToLetters.end()) {
                    return false;
                }

                auto insertTmp = std::make_pair(s[i], t[i]);
                letterMap.insert(insertTmp);
                mappedToLetters.insert(t[i]);
                continue;
            }
            else {
                if (itMap->second == t[i]) {
                    continue;
                }
                else {
                    return false;
                }
            }
        }

        return true;
    }
};